### PR TITLE
Adds a preference to have an intent check for guns.

### PR DESCRIPTION
**Pull request by chaoko99**
Title.

### CODE REVIEW

**Comment by printer16** [comment]
I disagree with this unless it's on a preference.

**Comment by skull132** [comment]
Preference or bust.

**Comment by chaoko99** [comment]
I can make it a pref toggle.

**Comment by skull132** [comment]
> I can make it a pref toggle.

Ye that should be fine. Default it to off as well.

**Comment by chaoko99** [comment]
Changed.

**Comment by OneOneThreeEight** [comment]
what's the point of adding it all then? Truly robust individuals have enough discipline to not have their gun in their active hand when not using it.

**Comment by chaoko99** [comment]
Is there a reason *not* to?

**Comment by OneOneThreeEight** [comment]
Because it's a gameplay preference that's going to be off by default for the people that actually need it but won't put in the effort to do so anyway?

If they're incompetent enough to let a gun go off and potentially hit someone because of an awful misclick then odds are they're not going to recognize the feature and use it. A very small minority of people would actually use this, and said people would probably end up learning not to wield a gun in-active-hand when not actively needing to use it.

**Comment by chaoko99** [comment]
All those thumbs down (Which shouldn't have any baring on anything, by the
way... x amount of people do not represent the entire community.) Are from
when I started this, not being optional, but forced on everyone. Now it's a
preference that defaults off. *Quit bitching about things that don't effect
you.*

**Comment by skull132** [comment]
Theoretically if we wanted to ration preference bitflags. But we have an
entirely new field to play with, so it's not necessary.

**Comment by OneOneThreeEight** [comment]
>How is this different than at
*least* 30% of this game's features?

Citation bloody needed. We removed the help intent forced onto guns on purpose because it served no purpose if people knew how to handle guns properly. Your average greytider that joins as security, Surprise! -- does not know how to properly utilize guns and compose themselves. They don't want to compose themselves, they want to shoot guns. Literally anyone else that is here either for the antags or for the roleplay knows already how not to cause misfires. There is a disproportionate amount of people that misfire because they're woefully incompetent.

>Also fun fact: you cannot change hands when holding a gun in two hands.
Another fun fact. You cannot put things in your off hand on tables. You
have to have them in your active hand.

_So unwield it or put it away until you need it._ Or people can just have a little discipline and this change is utterly unnecessary. Do I really need to hold your hand here by presenting these pretty clear solutions that any player could be using if they cared enough without the need for this code change?

>Jesus christ, I shouldn't have to justify a basic fucking feature that
doesn't hurt anyone unless they use it, and even then, in the most minor of
ways,

Yes, I'm pretty sure everyone is required to justify the purpose of every single suggestion or feature being pushed to implement. I have never seen another contributor as upset about this before.

>**I need this feature, fun fact, as I've
had trouble in the past in scenarios relating to this.** 

Okay, if it wasn't obvious before what the intention was, it's clear now.

>How about I go remove some aliens, because those people are putting
themselves at a disadvantage for playing them and are thusly not necessary.

Because aliens have literally nothing to do with a suggestion being functionally useless for an overwhelming majority of the playerbase that either doesn't care to use guns or simply doesn't recklessly use them with no consideration for other people. It's a vestigial feature. It serves no purpose besides for a tiny minority that knows about it, you, apparently, so you feel the need to enforce some degree of self-control **not by actually changing how _you_ play the game** but by **forcing the game to apply the handicap for you.**

File: code/modules/client/preferences.dm [hunk 1]
```diff
@@ -28,6 +28,7 @@ datum/preferences
 	var/UI_style_alpha = 255
 	var/motd_hash = ""					//Hashes for the new server greeting window.
 	var/memo_hash = ""
+	var/safety_check = FALSE //Checks if you are on non-harm when firing. Defaults off.
```
Comment (Lohikar) [inline code/modules/client/preferences.dm]: This change appears to be pointless with the swap to `parallax_togs`.

Comment (chaoko99) [reply code/modules/client/preferences.dm]: Shit shit shit

**Comment by chaoko99** [comment]
You are really, *really* going out of your way to have problems with things, aren't you?

**Comment by skull132** [comment]
Also, @OneOneThreeEight could you simmer?

It's a preference and we have some 10 more flags to use after this. It doesn't remove much and adds something that a few people might find helpful, specially if they're coming in from other Bay based 
servers. So I don't see the reason for fire and brimstone here.

**Comment by chaoko99** [comment]
Gonna be a hot minute. Been busy.

**Comment by OneOneThreeEight** [comment]
on second thought its not that bad.

though there are a couple of null refs. SQL code makes everyone cry.

**Comment by chaoko99** [comment]
Mmm. Me too.

File: SQL/migrate/V010__Preferences_secondary_toggles.sql [hunk 2]
```diff
@@ -0,0 +1,7 @@
+--
+-- Implemented in PR #3653.
+-- Renames parallax toggles to a more generic name.
+--
+
+ALTER TABLE `ss13_characters`
+  CHANGE `parallax_togs` `toggles_secondary` INT(11) NULL DEFAULT NULL;
```
Comment (skull132) [inline SQL/migrate/V010__Preferences_secondary_toggles.sql]: woups this is my bad it should be ``parallax_toggles`` instead of ``parallax_togs``

Comment (chaoko99) [reply SQL/migrate/V010__Preferences_secondary_toggles.sql]: Whoopsies

Comment (chaoko99) [reply SQL/migrate/V010__Preferences_secondary_toggles.sql]: Fixed.

File: code/modules/projectiles/targeting/targeting_triggers.dm [hunk 3]
```diff
@@ -9,6 +9,10 @@
 				AO.update_aiming_deferred()
 
 /obj/aiming_overlay/proc/trigger(var/perm)
+
+	if((user.client.prefs.parallax_togs & SAFETY_CHECK) && user.a_intent != I_HURT) //Check this first to save time.
```
[the remark this answers is not in the document]
Comment (chaoko99) [reply code/modules/projectiles/targeting/targeting_triggers.dm]: resolved

File: code/modules/client/preferences_toggles.dm [hunk 4]
```diff
@@ -214,3 +215,13 @@
...
+
+/client/verb/toggle_safety_check()
+
+	set name = "Toggle Gun Safety Check"
+	set category = "Preferences"
+	set desc = "Toggles checking for harm intent when firing."
+
+	prefs.parallax_togs ^= SAFETY_CHECK //Held in Parallax because we don't want to deal with an SQL migration right now.
+	prefs.save_preferences()
+	src << "Your guns will [(prefs.parallax_togs & SAFETY_CHECK) ? "no longer" : "now"] fire on non-help intents."
```
[the remark this answers is not in the document]
Comment (chaoko99) [reply code/modules/client/preferences_toggles.dm]: resolved

File: code/modules/client/preferences.dm (hunk 1 above)
Comment (chaoko99) [reply code/modules/client/preferences.dm]: resolved

**Comment by chaoko99** [comment]
All issues resolved. Tested and working.

**Comment by chaoko99** [comment]
>it fails travis

File: code/modules/client/preference_setup/global/02_settings.dm [hunk 5]
```diff
@@ -10,7 +10,7 @@
 	S["motd_hash"]        >> pref.motd_hash
 	S["memo_hash"]        >> pref.memo_hash
 	S["parallax_speed"]   >> pref.parallax_speed
-	S["parallax_toggles"] >> pref.parallax_togs
+	S["parallax_toggles"] >> pref.toggles_secondary
```
Comment (skull132) [inline code/modules/client/preference_setup/global/02_settings.dm]: This key should be renamed to `toggles_secondary`.

File: code/modules/client/preference_setup/global/02_settings.dm [hunk 6]
```diff
@@ -20,7 +20,7 @@
 	S["motd_hash"]        << pref.motd_hash
 	S["memo_hash"]        << pref.memo_hash
 	S["parallax_speed"]   << pref.parallax_speed
-	S["parallax_toggles"] << pref.parallax_togs
+	S["parallax_toggles"] << pref.toggles_secondary
```
Comment (skull132) [inline code/modules/client/preference_setup/global/02_settings.dm]: Ditto.

File: code/modules/client/preference_setup/global/02_settings.dm [hunk 7]
```diff
@@ -32,9 +32,9 @@
 				"asfx_togs",
 				"lastmotd" = "motd_hash",
 				"lastmemo" = "memo_hash",
-				"parallax_toggles" = "parallax_togs",
+				"parallax_toggles" = "toggles_secondary",
```
Comment (skull132) [inline code/modules/client/preference_setup/global/02_settings.dm]: Replace this line with `"toggles_secondary",` there is no need for the key-value pair, as both the column name and variable name will now line up.

File: code/modules/client/preference_setup/global/02_settings.dm [hunk 8]
```diff
@@ -66,7 +66,7 @@
 		"asfx_togs" = pref.asfx_togs,
 		"lastmotd" = pref.motd_hash,
 		"lastmemo" = pref.memo_hash,
-		"parallax_toggles" = pref.parallax_togs,
+		"parallax_toggles" = pref.toggles_secondary,
```
Comment (skull132) [inline code/modules/client/preference_setup/global/02_settings.dm]: Rename `"parallax_toggles"` to `"toggles_secondary"`, keep the value of it tho here.

File: code/modules/client/preferences_toggles.dm [hunk 9]
```diff
@@ -207,10 +208,20 @@
...
+
+/client/verb/toggle_safety_check()
+
+	set name = "Toggle Gun Safety Check"
+	set category = "Preferences"
+	set desc = "Toggles firing guns on intents other than help."
+
+	prefs.toggles_secondary ^= SAFETY_CHECK //Held in Parallax because we don't want to deal with an SQL migration right now.
+	prefs.save_preferences()
+	src << "You will [(prefs.toggles_secondary & SAFETY_CHECK) ? "no longer" : "now"] fire your weapon on intents other than help."
```
Comment (Arrow768) [inline code/modules/client/preferences_toggles.dm]: Shouldnt this be hurt instead of help ?

Comment (OneOneThreeEight) [reply code/modules/client/preferences_toggles.dm]: << other than >> help is grammatically correct.

**Comment by chaoko99** [comment]
They are right though.

File: code/modules/client/preference_setup/global/02_settings.dm (hunk 8 above)
Comment (chaoko99) [reply code/modules/client/preference_setup/global/02_settings.dm]: Resolved.

File: code/modules/client/preference_setup/global/02_settings.dm (hunk 6 above)
Comment (chaoko99) [reply code/modules/client/preference_setup/global/02_settings.dm]: Resolved.

File: code/modules/client/preference_setup/global/02_settings.dm (hunk 5 above)
Comment (chaoko99) [reply code/modules/client/preference_setup/global/02_settings.dm]: Resolved.

File: code/modules/projectiles/gun.dm [hunk 10]
```diff
@@ -151,9 +151,14 @@
 		return ..() //Pistolwhippin'
 
 /obj/item/weapon/gun/proc/Fire(atom/target, mob/living/user, clickparams, pointblank=0, reflex=0)
+
+
 	if(!user || !target) return
 
 	add_fingerprint(user)
+	if((user.client.prefs.toggles_secondary & SAFETY_CHECK) && user.a_intent != I_HURT) //Check this first to save time.
```
Comment (skull132) [inline code/modules/projectiles/gun.dm]: I realized, it is also possible that user doesn't have a client. Poke a `user.client` into this if chain here.

So: `if (user.client && (user.client.prefs.toggles_secondary & SAFETY_CHECK) && user.a_intent != I_HURT)`

Comment (chaoko99) [reply code/modules/projectiles/gun.dm]: Resolved.

**Comment by chaoko99** [comment]
Merge me daddy.

**Comment by chaoko99** [comment]
Still feature-complete.

File: code/modules/client/preference_setup/global/02_settings.dm (hunk 5 above)
Comment (skull132) [reply code/modules/client/preference_setup/global/02_settings.dm]: Noot yet. The `S["parallax_toggles"]` should be `S["toggles_secondary"]` like in saving.

**Comment by chaoko99** [comment]
Resolved. Merge me daddy.

**Pull request merged**